### PR TITLE
Provide a spellchecking suggestion when someone mispells a constructor.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/28244
Fixes https://github.com/dotnet/roslyn/issues/24997

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @jcouv @jinujoseph

File: src/EditorFeatures/CSharpTest/Diagnostics/SpellCheck/SpellCheckTests.cs [hunk 1]
```diff
@@ -576,6 +576,21 @@ void M()
 }",
 @"class C<T> where T : unmanaged
 {
+}");
+        }
+
+        [WorkItem(1688, "https://github.com/dotnet/csharplang/issues/1688")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsSpellcheck)]
+        public async Task TestMispelledConstructor()
```
Comment (KirillOsenkov) [inline src/EditorFeatures/CSharpTest/Diagnostics/SpellCheck/SpellCheckTests.cs]: You misspelled the word `Mispelled`

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/Diagnostics/SpellCheck/SpellCheckTests.cs]: Tnx :)

File: src/Features/Core/Portable/SpellCheck/AbstractSpellCheckCodeFixProvider.cs [hunk 2]
```diff
@@ -29,11 +35,23 @@ internal abstract class AbstractSpellCheckCodeFixProvider<TSimpleName> : CodeFix
 
             var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
             var node = syntaxRoot.FindNode(span);
```
Comment (jcouv) [inline src/Features/Core/Portable/SpellCheck/AbstractSpellCheckCodeFixProvider.cs]: >FindNode [](start = 34, length = 8)

I didn't follow the "node or token" logic. When would we not find a node?

**Comment by jcouv** [comment]
It looks like there's a build error:
`D:\j\workspace\windows_debug---09a07fba\src\EditorFeatures\VisualBasicTest\Diagnostics\Spellcheck\SpellcheckTests.vb(8,9): error BC40056: Namespace or type specified in the Imports 'Microsoft.CodeAnalysis.VisualBasic.CodeFixes.Spellcheck' doesn't contain any public member or cannot be found. Make sure the namespace or the type is defined and contains at least one public member. Make sure the imported element name doesn't use any aliases. [D:\j\workspace\windows_debug---09a07fba\src\EditorFeatures\VisualBasicTest\BasicEditorServicesTest.vbproj]`

File: src/Features/Core/Portable/SpellCheck/AbstractSpellCheckCodeFixProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/SpellCheck/AbstractSpellCheckCodeFixProvider.cs]: @jcouv when you have a ConstructorDeclaration, the 'name' is not an IdentifierNameSyntax, it's just an 'Identifier Token'.

**Comment by CyrusNajmabadi** [comment]
retest this please.

File: src/EditorFeatures/VisualBasicTest/Diagnostics/Spellcheck/SpellcheckTests.vb [hunk 3]
```diff
@@ -5,7 +5,7 @@ Imports Microsoft.CodeAnalysis.CodeActions
 Imports Microsoft.CodeAnalysis.CodeFixes
 Imports Microsoft.CodeAnalysis.Diagnostics
 Imports Microsoft.CodeAnalysis.Editor.UnitTests.Extensions
-Imports Microsoft.CodeAnalysis.VisualBasic.CodeFixes.Spellcheck
+Imports Microsoft.CodeAnalysis.VisualBasic.SpellCheck
 Imports Microsoft.CodeAnalysis.VisualBasic.Diagnostics
 
 Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.Diagnostics.Spellcheck
```
Comment (mavasani) [inline src/EditorFeatures/VisualBasicTest/Diagnostics/Spellcheck/SpellcheckTests.vb]: Can you please add a VB spell check test as well?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/VisualBasicTest/Diagnostics/Spellcheck/SpellcheckTests.vb]: No :)  VB doesn't have named constructors...

**Comment by jcouv** [comment]
@jinujoseph for ask-mode approval for 15.8. Thanks

**Comment by jinujoseph** [comment]
Approved to merge for 15.8.Preview5 (Once test green)

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
@jinujoseph required tests pass.  I don't know what the story is on integration tests.  My success rate with them seems to be near 25%.

**Comment by CyrusNajmabadi** [comment]
Looking at the queues, it looks like integration tests are just exceptionally flakey.

**Comment by mavasani** [comment]
Ignoring the unrelated flaky [integration test failure](https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_release_vs-integration_prtest/2545/)

**Comment by CyrusNajmabadi** [comment]
Thanks all!

**Pull request merged**